File: java/yass/test/ch/softappeal/yass/transport/test/MessageSerializerTest.java
package ch.softappeal.yass.transport.test;

import ch.softappeal.yass.remote.ExceptionReply;
import ch.softappeal.yass.remote.Message;
import ch.softappeal.yass.remote.Request;
import ch.softappeal.yass.remote.ValueReply;
import ch.softappeal.yass.serialize.JavaSerializer;
import ch.softappeal.yass.serialize.Serializer;
import ch.softappeal.yass.serialize.test.JavaSerializerTest;
import ch.softappeal.yass.transport.MessageSerializer;
import org.junit.Assert;
import org.junit.Test;

import java.io.EOFException;
import java.util.List;

public class MessageSerializerTest {

    private static final Serializer SERIALIZER = new MessageSerializer(JavaSerializer.INSTANCE);

    private static <T extends Message> T copy(final T value) throws Exception {
        return JavaSerializerTest.copy(SERIALIZER, value);
    }

    @Test public void request() throws Exception {
        final var serviceId = 123456;
        final var methodId = 4711;
        final var request = copy(new Request(serviceId, methodId, List.of()));
        Assert.assertEquals(serviceId, request.serviceId);
        Assert.assertEquals(methodId, request.methodId);
        Assert.assertTrue(request.arguments.isEmpty());
    }

    @Test public void value() throws Exception {
        final var value = "xyz";
        final var reply = copy(new ValueReply(value));
        Assert.assertEquals(value, reply.value);
    }

    @Test public void exception() throws Exception {
        final var reply = copy(new ExceptionReply(new EOFException()));
        Assert.assertTrue(reply.exception instanceof EOFException);
    }

}

File: kotlin/yass/test/ch/softappeal/yass/transport/MessageSerializerTest.kt
package ch.softappeal.yass.transport

import ch.softappeal.yass.remote.*
import ch.softappeal.yass.serialize.*
import java.io.*
import kotlin.test.*

private val SERIALIZER = messageSerializer(JavaSerializer)

private fun <T : Message?> copy(value: T): T = copy(SERIALIZER, value)

class MessageSerializerTest {

    @Test
    fun request() {
        val serviceId = 123456
        val methodId = 4711
        val request = copy(Request(serviceId, methodId, listOf()))
        assertEquals(serviceId, request.serviceId)
        assertEquals(methodId, request.methodId)
        assertTrue(request.arguments.isEmpty())
    }

    @Test
    fun value() {
        val value = "xyz"
        val reply = copy(ValueReply(value))
        assertEquals(value, reply.value)
    }

    @Test
    fun exception() {
        val reply = copy(ExceptionReply(EOFException()))
        assertTrue(reply.exception is EOFException)
    }

    @Test
    fun write1() = assertEquals(
        "unexpected value 'null'",
        assertFailsWith<IllegalStateException> {
            SERIALIZER.write(writer(ByteArrayOutputStream()), null)
        }.message
    )

    @Test
    fun write2() = assertEquals(
        "unexpected value 'error'",
        assertFailsWith<IllegalStateException> {
            SERIALIZER.write(writer(ByteArrayOutputStream()), "error")
        }.message
    )

    @Test
    fun read() = assertEquals(
        "unexpected type 123",
        assertFailsWith<IllegalStateException> {
            SERIALIZER.read(reader(ByteArrayInputStream(byteArrayOf(123))))
        }.message
    )
}
